### PR TITLE
dispatch_json core, attempt 2

### PR DESCRIPTION
**Pull request by bartlomieju**

### CODE REVIEW

File: cli/ops/fs.rs [hunk 1]
```diff
@@ -146,6 +150,94 @@ fn op_open(
...
+
+use std::pin::Pin;
+use futures::Future;
+
+fn op_open_async(
+  isolate_state: &mut CoreIsolateState,
+  state: &State,
+  args: Value,
+  _zero_copy: &mut [ZeroCopyBuf],
+) -> Pin<Box<dyn Future<Output = Result<Value, OpError>>>> {
```
Comment (bartlomieju) [inline cli/ops/fs.rs:220]: @piscisaureus could you take a look at `stateful_json_op_async` to turn this function into `async fn`?

Comment (bartlomieju) [reply cli/ops/fs.rs:220]: I've tried this:
```rust
async fn op_open_async(
  isolate_state: &mut CoreIsolateState,
  state: &State,
  args: Value,
  _zero_copy: &mut [ZeroCopyBuf],
) -> Result<Value, OpError> {
  let resource_table = isolate_state.resource_table.clone();
  let (path, open_options) = open_helper(&state, args)?;
  let tokio_file = tokio::fs::OpenOptions::from(open_options)
    .open(path)
    .await?;
  let mut resource_table = resource_table.borrow_mut();
  let rid = resource_table.add(
    "fsFile",
    Box::new(StreamResourceHolder::new(StreamResource::FsFile(Some((
      tokio_file,
      FileMetadata::default(),
    ))))),
  );
  Ok(json!(rid))
}
```
```rust
pub fn stateful_json_op_async<D, F>(&self, dispatcher: D) -> impl OpDispatcher
  where
    D: Fn(
      &mut deno_core::CoreIsolateState,
      &State,
      Value,
      &mut [ZeroCopyBuf],
    ) -> F,
    F: Future<Output = Result<Value, OpError>> + 'static
  {
  ... 
}
```
but I'm getting this error trying to compile:
```
▶ cargo build
   Compiling deno v1.2.3 (/Users/biwanczuk/dev/deno/cli)
error[E0271]: type mismatch resolving `for<'r, 's, 't0> <for<'_, '_, '_> fn(&mut deno_core::CoreIsolateState, &state::State, serde_json::Value, &mut [deno_core::ZeroCopyBuf]) -> impl futures::Future {ops::fs::op_open_async} as std::ops::FnOnce<(&'r mut deno_core::CoreIsolateState, &'s state::State, serde_json::Value, &'t0 mut [deno_core::ZeroCopyBuf])>>::Output == _`
  --> cli/ops/fs.rs:26:36
   |
26 |   i.register_op("op_open_async", s.stateful_json_op_async(op_open_async));
   |                                    ^^^^^^^^^^^^^^^^^^^^^^ expected bound lifetime parameter, found concrete lifetime

error: aborting due to previous error

For more information about this error, try `rustc --explain E0271`.
error: could not compile `deno`.
```

**Comment by piscisaureus** [comment]
@bartek + @ry 

So I got it to work I believe.

Ultimately this was solved not by significantly changing the signature of `stateful_json_op_async()`, but rather by changing:

```rust
async fn op_open_async(
  isolate_state: &mut CoreIsolateState, state: &State, args: Value, _zero_copy: &mut [ZeroCopyBuf]
) -> Result<Value, OpError>> {
  // ...
}
```

to

```rust
fn op_open_async(
  isolate_state: &mut CoreIsolateState, state: &State, args: Value, _zero_copy: &mut [ZeroCopyBuf]
) -> impl futures::future::Future<Output = Result<Value, OpError>> {
  async move {
    // ...
  }
}
```

Apparently when you have an async function that borrows something (e.g. `async fn my_async(&something) {}`) the borrow needs to be valid for the static lifetime (it's equivalent to `fn my_async(&'static something) -> impl Future<Output = ()> {}`), which explains why I had trouble capturing `&State` in the closure.

**Comment by piscisaureus** [comment]
Here's the theory to go with it: 
https://rust-lang.github.io/async-book/03_async_await/01_chapter.html#async-lifetimes

**Comment by bartlomieju** [comment]
@ry @piscisaureus I think given that we can't use `async fn` I propose we try one more solution, ie. each op dispatcher returns a variant of:
```rust
enum JsonOp {
  Sync(Result<Value, Error>),
  Async(impl Future<Output = Result<Value, Error>>),
  ...
}
```

which would make dispatchers look like:

```rust
fn op_open_sync(
  isolate_state: &mut CoreIsolateState, state: &State, args: Value, _zero_copy: &mut [ZeroCopyBuf]
) -> JsonOp {
  JsonOp::Sync({
  ...
  })
}

fn op_open_async(
  isolate_state: &mut CoreIsolateState, state: &State, args: Value, _zero_copy: &mut [ZeroCopyBuf]
) -> JsonOp {
  JsonOp::Async(async move {
  ...
  })
}
```
We already use similar structure for `MinimalOp`. WDYT?

**Pull request closed without merging**